portal/models/qb_timeline.py

```diff
@@ -736,6 +762,77 @@ def invalidate_users_QBT(user_id, research_study_id):
     db.session.commit()
 
 
+def check_for_overlaps(qbt_rows, cli_presentation=False):
+    """Sanity function to confirm users timeline doesn't contain overlaps"""
+    # Expect ordered rows with increasing `at` values.  Track (QB,iterations)
+    # seen, notify if overlaps are discovered.
+    from .questionnaire_bank import QuestionnaireBank
+
+    def rp_name_from_qb_id(qb_id):
+        return ResearchProtocol.query.join(QuestionnaireBank).filter(
+            QuestionnaireBank.research_protocol_id ==
+            ResearchProtocol.id).filter(
+            QuestionnaireBank.id == qb_id).with_entities(
+            ResearchProtocol.name).first()[0]
+
+    def int_or_none(value):
+        """None safe int cast from string"""
+        if value is None or value == 'None':
+            return None
+        return int(value)
+
+    seen = set()
+    last_at, previous_key = None, None
+    reported_on = set()
+    for row in qbt_rows:
+        # Confirm expected order
+        if last_at:
+            assert row.at >= last_at
+
+        key = f"{row.qb_id}:{row.qb_iteration}"
+        if previous_key and previous_key != key:
+            # just moved to next visit, confirm it's novel
+            if key in seen:
+                if not (key in reported_on and previous_key in reported_on):
```

Comment:
Minor issue-some of these could be inverted into guard clauses (with `continue`) to make it less indented

Reply:
not easy, given the variables set following that clause.  punting for now, but thanks for the suggestion, i like your eye toward better!